File: src/viser/client/_bench/run-bench.mjs
import { chromium } from "playwright";
import { fileURLToPath } from "node:url";
import path from "node:path";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const htmlPath = path.join(__dirname, "bench.html");

const browser = await chromium.launch({
  headless: false, // headed -> real GPU (ANGLE/Metal) on macOS
  args: ["--ignore-gpu-blocklist", "--enable-gpu-rasterization"],
});
const page = await browser.newPage();
page.on("console", (m) => console.error("[page]", m.text()));
page.on("pageerror", (e) => console.error("[pageerror]", e.message));

await page.goto("file://" + htmlPath);
await page.waitForFunction("window.__ready === true", { timeout: 15000 });
const results = await page.evaluate(() => window.bench());
console.log(JSON.stringify(results, null, 2));
await browser.close();

File: src/viser/client/_bench/run_bench.py
import json

from playwright.sync_api import sync_playwright

html = "http://localhost:8899/_bench/bench.html"

with sync_playwright() as p:
    browser = p.chromium.launch(
        headless=False,  # headed -> real GPU on macOS
        args=[
            "--use-angle=metal",
            "--ignore-gpu-blocklist",
            "--enable-gpu",
            "--disable-gpu-sandbox",
        ],
    )
    page = browser.new_page()
    page.on("console", lambda m: print("[page]", m.text))
    page.on("pageerror", lambda e: print("[pageerror]", e))
    page.goto(html)
    page.wait_for_function("window.__ready === true", timeout=15000)
    results = page.evaluate("() => window.bench()")
    print("RESULTS_JSON_START")
    print(json.dumps(results, indent=2))
    print("RESULTS_JSON_END")
    browser.close()
